### PR TITLE
Add support for snapshot and repeated plan

### PR DESCRIPTION
**Pull request by dkeven**
Signed-off-by: dkeven <keven@kubesphere.io>

This PR adds:

a `jibu-backup-method` flag to support csi snapshot copy method, in addition to the existing `filesystem`

a `jibu-backup-repeat-enabled` flag to support repeated backup plan, as well as a  `jibu-backup-frequency` flag to specify the cronjob str, and a `jibu-bakcup-repeat-check-num` flag to specify how many times to check the repeatedly created backup jobs.

after the test, all of the backup jobs and restore jobs will be deleted.

if the backup plan is repeated, it will also get deleted.

### CODE REVIEW

File: test/jibu/args.go [hunk 1]
```diff
@@ -3,38 +3,60 @@ package jibu
...
-	argSkipRestore = flag.Bool("jibu-skip-restore", false, "if set, skip restore test")
-	argBackupCluster = flag.String("jibu-backup-cluster", "", "if set, use specified cluster for backup")
-	argRestoreCluster = flag.String("jibu-restore-cluster", "", "if set, use specified cluster for restore")
-	argStorage = flag.String("jibu-storage", "", "if set, use specified storage during test")
-	argCleanUpOnEnd = flag.Bool("jibu-clean-up-on-end", true, "clean up after test, including delete backup/restore jobs, delete restored namespace etc.")
-	argBackupPlanName = flag.String("jibu-backup-plan-name", "", "backup plan name, if not set, will use backup-{timestamp}")
-	argBackupJobName = flag.String("jibu-backup-job-name", "", "backup job name, if not set, will use {backup-plan-name}-{random-string}")
-	argRestorePlanName = flag.String("jibu-restore-plan-name", "", "restore plan name, if not set, will use restore-{timestamp}")
-	argRestoreJobName = flag.String("jibu-restore-job-name", "", "restore job name, if not set, will use {restore-plan-name}-{random-string}")
+	argBackupRepeatEnabled    = flag.Bool("jibu-backup-repeat-enabled", false, "wether create a repeted backupplan")
```
Comment (stoneshi-yunify) [inline test/jibu/args.go]: typo: whether

File: test/jibu/args.go [hunk 2]
```diff
@@ -3,38 +3,60 @@ package jibu
...
+	argBackupFrequency        = flag.String("jibu-backup-frequency", "*/3 * * * *", "the frequency(crontab string) to create backup jobs, defaults to every 3 minutes for faster testing, only effetive when backup-repeat-enabled is set to true")
```
Comment (stoneshi-yunify) [inline test/jibu/args.go]: typo: effective

File: test/jibu/backup_test.go [hunk 3]
```diff
@@ -147,32 +160,62 @@ var _ = Describe("use jibu api", func() {
...
+						indexChan <- i
+						wg.Add(1)
+					}
+					checkRepeatedBackupJob := func() {
+						defer wg.Done()
+						index := <-indexChan
+						MyBy(fmt.Sprintf("wait for backup job to be created in %v, index: %d", backupJobRepeatedCreationTimeout, index))
+						jobName := waitNthBackupJobCreation(jibuClient, tenant, backupPlanName, index)
+						MyBy(fmt.Sprintf("backup job created, index: %d, name: %s", index, jobName))
+						MyBy(fmt.Sprintf("backup job should complete in %v, index: %d, name: %s", backupJobFinishedTimeout, index, jobName))
```
Comment (stoneshi-yunify) [inline test/jibu/backup_test.go:220]: question: if every 3 minutes generate a backup job, but each backup job needs take 1 hour to complete, what would happen? will the second backup job be generated 3 minutes later or 1 hour later?

File: test/jibu/args.go [hunk 4]
```diff
@@ -3,38 +3,60 @@ package jibu
...
-	jobRetention              = 240
-	backupPlanReadyTimeout    = 5 * time.Minute
-	backupJobFinishedTimeout  = 2 * time.Hour
-	restorePlanReadyTimeout   = 5 * time.Minute
-	restoreJobFinishedTimeout = 2 * time.Hour
-	actionStartJob            = "StartJob"
+	resourcePickRetryLimit           = 10
+	jobRetention                     = 240
+	backupPlanReadyTimeout           = 5 * time.Minute
+	backupJobRepeatedCreationTimeout = 3 * time.Minute
```
Comment (stoneshi-yunify) [inline test/jibu/args.go:16]: why set to 3 minutes? is this somehow related to the default frequency?

File: test/jibu/backup_test.go [hunk 5]
```diff
@@ -75,23 +87,21 @@ var _ = Describe("use jibu api", func() {
...
-					MyBy(spew.Sdump("backupjob", backupJob))
-				}
+				_, _, _ = jibuClient.BackupPlanTagApi.DeleteBackupPlan(ctx, tenant, backupPlanName)
+				_ = deleteJobsOfBackupPlan(jibuClient, tenant, backupPlanName)
 				restoreJob, _, err := jibuClient.RestoreJobTagApi.GetRestoreJob(ctx, tenant, restoreJobName)
 				if err == nil {
 					MyBy(spew.Sdump("restorejob", restoreJob))
 				}
-				_, _, _ = jibuClient.BackupJobTagApi.DeleteBackupJob(ctx, tenant, backupJobName)
+				_, _, _ = jibuClient.RestorePlanTagApi.DeleteRestorePlan(ctx, tenant, restorePlanName)
```
Comment (stoneshi-yunify) [inline test/jibu/backup_test.go]: why delete restore plan? I think it has no harm to later testings.

**Review by stoneshi-yunify — CHANGES_REQUESTED** [review]

File: test/jibu/backup_test.go (hunk 3 above)
Comment (dkeven) [reply test/jibu/backup_test.go:220]: After some real experiments,  it turns out that the second backup job will only be created *after* the first backup job is compeleted.

That said, if a backup job stucks, then no backup job will be created as the configured cron schedule.

File: test/jibu/args.go (hunk 4 above)
Comment (dkeven) [reply test/jibu/args.go:16]: No, it has no relation with the default frequency, just a toleration of delay to wait for the scheduled backup job to be created, because I thought the backup job might not be created precisely on time of the cron schedule, as there might be some pre-processing/delay, etc.

File: test/jibu/backup_test.go (hunk 5 above)
Comment (dkeven) [reply test/jibu/backup_test.go]: Just thought the plan is no longer needed. If this deletion is not wanted, I'll change it.

File: test/jibu/args.go (hunk 4 above)
Comment (stoneshi-yunify) [reply test/jibu/args.go:16]: okay

File: test/jibu/backup_test.go (hunk 5 above)
Comment (stoneshi-yunify) [reply test/jibu/backup_test.go]: the general idea is to keep some data/clues in case we want to debug after a long-run test. Deleting it is fine, but may need print it in log before delete it.

File: test/jibu/backup_test.go (hunk 3 above)
Comment (dkeven) [reply test/jibu/backup_test.go:220]: So now I think the current implementation of cron-scheduled checking might not work if the backup job takes a long time to finish: if the first backup job has not finished yet, but the second checking cron job is triggered, it will timeout waiting for the second backup job to be created.

I'll change it to fit into this scenario.

File: test/jibu/backup_test.go (hunk 5 above)
Comment (dkeven) [reply test/jibu/backup_test.go]: Okay, I'll leave the restorePlan and delete the backupPlan(with a spew before) if it's a repeated one, since the latter will generate more backupjobs.

**Review by stoneshi-yunify — APPROVED** [review]
